File: week-4/Java/LinkedList.java
class LinkedList { 
  
    //static Node head; 
  
    static class Node { 
  
        char data; 
        Node next; 
  
        Node(char d) 
        { 
            data = d; 
            next = null; 
        } 
    } 
  
    public static Node reverseLL(Node a){

        StringBuilder temp = new StringBuilder();

        while(a.next != null){
            temp.append(a.data);
        }

        temp.reverse();
        String s = temp.toString();
        Node t = new Node ();
        Node head = t;

        for(int i = 0;i<s.length();i++){
            t.data = s.charAt(i);
            t = t.next();
        }

        return head;

    }

    public static Node createList(String s){
        Node list =  new Node();
        Node list2 = list;
        for (int i = 0;i<s.length();i++){
            list.data = s.charAt(i);
            list = list.next;
        }
        return list2;
    }

    public static printList(Node a){
        while(a.data){
            System.out.println(a.data);
            a = a.next;
        }
    }
    public static void main(String[] args) 
    { 
        String s = "hello";
        //System.out.println("hi");
        Node list = createList(s);
        list = reverseLL(list);
        printList(list);      

      
    } 
} 

File: week-4/C++/Robert-Robinson/LinkedList.cpp
#include "LinkedList.h"
using namespace std;

template <typename T>
void LinkedList<T>::insert(T data){
	Node<T> * temp = new Node<T>;
	temp->data = data;
	temp->next=nullptr;
	num++;

	if(head == nullptr){
		head = curr = tail = temp;
		temp->prev = nullptr;
		return;
	}
	
	tail->next=temp;
	temp->prev=tail;
	tail=temp;
}

template <typename T>
int LinkedList<T>::next(){
	if(curr==tail){
		//printf("no more elements");
		return 0;
	}
	curr = curr->next;
	return 1;
}

template <typename T>
T LinkedList<T>::getCurr(){
	return curr->data;
}

template <typename T>
int LinkedList<T>::getNum(){
	return this->num;
}

template <typename T>
void LinkedList<T>::reset(){
	curr=head;
}

template <typename T>
bool LinkedList<T>::isMore(){
	if(curr->next!=nullptr){
		return true;
	}
	return false;
}

template <typename T>
void LinkedList<T>::print(){
	Node<T> * temp = head;
	while(temp!=nullptr){
		printf("%d ", temp->data);
	//	cout << temp->data + " ";
		temp = temp->next;
	}
	//cout << "\n";
	printf("\n");
}

template <>
void LinkedList<string>::print(){
	Node<string> * temp = head;
	while(temp!=nullptr){
		//printf("%s ", temp->data);
		cout << temp->data + " ";
		temp = temp->next;
	}
	//cout << "\n";
	printf("\n");
}

template <>
string LinkedList<string>::pop(){
	this->reset(); //////////////////////////////////////////////////////////
	string ret=tail->data;
	Node<string> * temp;
	if(head==nullptr){
		return nullptr;
	}
	
	if(tail == head){
		//ret=tail->data;
		temp = tail;
		head=tail=curr=nullptr;
		delete temp;
		num--;
		return ret;
	}
	
	//ret=tail->data;
	temp = tail;
	tail=tail->prev;
	tail->next=nullptr;
	delete temp;
	num--;
	return ret;
}

template class LinkedList<int>;
template class LinkedList<string>;
